data/registry_model/registry_proxy_model.py

```diff
@@ -185,13 +191,36 @@ def lookup_manifest_by_digest(
                 oci.tag.set_tag_end_ms(db_tag, new_expiration)
                 # if the manifest is a child of a manifest list in this repo, renew
                 # the parent(s) manifest list tag too.
-                links = ManifestChild.select(ManifestChild.manifest_id).where(
-                    (ManifestChild.repository_id == repository_ref.id)
-                    & (ManifestChild.child_manifest_id == wrapped_manifest.id)
+                # select tag ids by most recent lifetime_end_ms uniquely by name,
+                # based on the link between sub-manifest and parent manifest in the
+                # manifest link.
+                subq = (
+                    TagTable.select(
+                        fn.row_number()
+                        .over(partition_by=TagTable.name, order_by=TagTable.lifetime_end_ms.desc())
+                        .alias("rnk"),
+                        TagTable.id,
+                        TagTable.name,
+                        TagTable.lifetime_end_ms,
+                    )
+                    .join(ManifestChild, on=(TagTable.manifest_id == ManifestChild.manifest_id))
+                    .where(
+                        (ManifestChild.repository_id == repository_ref.id)
+                        & (ManifestChild.child_manifest_id == wrapped_manifest.id)
+                    )
+                    .order_by(TagTable.name)
+                    .order_by_extend(TagTable.lifetime_end_ms.desc())
                 )
-                parent_ids = [link.manifest_id for link in links]
+                items = (
+                    Select(columns=[subq.c.rnk, subq.c.id, subq.c.name, subq.c.lifetime_end_ms])
+                    .from_(subq)
+                    .where(subq.c.rnk == 1)
+                    .bind(db)
+                )
+                tag_ids = [item["id"] for item in items]
```

Comment:
this query turned out way more complicated than I wanted it to be. I started using `DISTINCT ON` but sqlite doesn't support that (only `DISTINCT` without `ON`), so this is how I managed to make the same query work on both postgres and sqlite.

Reply:
ugh typical
https://github.com/quay/quay/runs/7790348396?check_suite_focus=true#step:5:1181